go/proxyd/daisychain.go

```diff
@@ -0,0 +1,691 @@
+		epoch1:      backends["epoch1"],
+		epoch2:      backends["epoch2"],
+		epoch3:      backends["epoch3"],
+		epoch4:      backends["epoch4"],
+		epoch5:      backends["epoch5"],
+		epoch6:      backends["epoch6"],
+		maxBodySize: maxBodySize,
+	}
+	return &srv
+}
+
+func StartDaisyChain(config *Config) (func(), error) {
+	if err := config.ValidateDaisyChainBackends(); err != nil {
+		return nil, err
+	}
+
+	// TODO: figure out how to not need to pass
+	// in the rate limiter here by parsing the
+	// args in the config and creating it in there
+	lim := NewLocalRateLimiter()
+	_, backendsByName, err := config.BuildBackends(lim)
+	if err != nil {
+		return nil, err
+	}
+
+	// parse the config
+	srv := NewDaisyChainServer(
+		backendsByName,
+		config.Server.MaxBodySizeBytes,
+	)
+
+	// send a chain id request to each node to ensure they are on the same chain
+	req, _ := ParseRPCReq([]byte(`{"id":"1","jsonrpc":"2.0","method":"eth_chainId","params":[]}`))
+	chainIds := []*hexutil.Big{}
+	for _, backend := range srv.Backends() {
+		res, _ := backend.Forward(context.Background(), req)
+		str, ok := res.Result.(string)
+		if !ok {
+			return nil, errors.New("cannot fetch chainid on start")
+		}
+		chainId := new(hexutil.Big)
+		chainId.UnmarshalText([]byte(str))
+		chainIds = append(chainIds, chainId)
+	}
+
+	if len(chainIds) == 0 {
+		panic("cannot fetch remote chain id")
+	}
+	chainId := chainIds[0].ToInt()
+	for _, id := range chainIds {
+		if id.ToInt().Cmp(chainId) != 0 {
+			panic("mismatched chain ids detected")
+		}
+	}
+	log.Info("detected chain id", "value", chainId)
+	srv.chainId = chainId
+
+	if config.Metrics.Enabled {
+		addr := fmt.Sprintf("%s:%d", config.Metrics.Host, config.Metrics.Port)
+		log.Info("starting metrics server", "addr", addr)
+		go http.ListenAndServe(addr, promhttp.Handler())
+	}
+
+	// To allow integration tests to cleanly come up, wait
+	// 10ms to give the below goroutines enough time to
+	// encounter an error creating their servers
+	errTimer := time.NewTimer(10 * time.Millisecond)
+
+	if config.Server.RPCPort != 0 {
+		go func() {
+			if err := srv.RPCListenAndServe(config.Server.RPCHost, config.Server.RPCPort); err != nil {
+				if errors.Is(err, http.ErrServerClosed) {
+					log.Info("RPC server shut down")
+					return
+				}
+				log.Crit("error starting RPC server", "err", err)
+			}
+		}()
+	}
+
+	if config.Server.WSPort != 0 {
+		go func() {
+			if err := srv.WSListenAndServe(config.Server.WSHost, config.Server.WSPort); err != nil {
+				if errors.Is(err, http.ErrServerClosed) {
+					log.Info("WS server shut down")
+					return
+				}
+				log.Crit("error starting WS server", "err", err)
+			}
+		}()
+	}
+
+	<-errTimer.C
+	log.Info("started daisychain")
+
+	return func() {
+		log.Info("shutting down daisychain")
+		srv.Shutdown()
+		log.Info("goodbye")
+	}, nil
+}
+
+// TODO: batch support
+func (s *DaisyChainServer) HandleRPC(w http.ResponseWriter, r *http.Request) {
+	ctx := s.populateContext(w, r)
```

Comment:
note to self: reduce code copy here by turning the beginning of handle rpc into a function that can be used by both services